File: Selenium/Activities/Activity21.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.firefox.service import Service as FirefoxService
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait
from webdriver_manager.firefox import GeckoDriverManager

service = FirefoxService(GeckoDriverManager().install())
with webdriver.Firefox(service=service) as driver:

    driver.get("https://v1.training-support.net/selenium/tab-opener")
    print("Page title is: ", driver.title)
    wait = WebDriverWait(driver, 10)

    driver.find_element(By.TAG_NAME,"a").click()
    wait.until(expected_conditions.number_of_windows_to_be(2))
    print("Number of handles: ",driver.window_handles)
    for handle in driver.window_handles:
        driver.switch_to.window(handle)

    #wait.until(expected_conditions.number_of_windows_to_be(2))
    print("Newly opened tab: ",driver.window_handles)
    wait.until(expected_conditions.number_of_windows_to_be(2))
    print("Page title is: ", driver.title)
    heading=driver.find_element(By.XPATH,"//div[@class='content']")
    print(heading.text)

    driver.find_element(By.ID,"actionButton").click()

    wait.until(expected_conditions.number_of_windows_to_be(3))
    #Switch focus
    for handle1 in driver.window_handles:
        driver.switch_to.window(handle1)

    wait.until(expected_conditions.number_of_windows_to_be(3))
    print("Current tab: ", driver.current_window_handle)
    print("Page title: ", driver.title)
    print(driver.find_element(By.CLASS_NAME,"content").text)
    driver.quit()

File: Selenium/Activities/Activity21.java
package Activities;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

public class Activity21 {
    public static void main(String[] args)  {
        WebDriverManager.firefoxdriver().setup();
        WebDriver driver = new FirefoxDriver();
        driver.get("https://v1.training-support.net/selenium/tab-opener");
        System.out.println("Title: " + driver.getTitle());
        driver.findElement(By.tagName("a")).click();
        WebDriverWait wait=new WebDriverWait(driver, Duration.ofSeconds(10));
        wait.until(ExpectedConditions.numberOfWindowsToBe(2));
        System.out.println("Number of handles: "+driver.getWindowHandles());
        for(String handle:driver.getWindowHandles()) {
            driver.switchTo().window(handle);
        }
        wait.until(ExpectedConditions.elementToBeClickable(By.id("actionButton")));
        System.out.println("Newly opened tab " + driver.getWindowHandle());
        System.out.println("Page title: " + driver.getTitle());
        System.out.println("Heading: "+ driver.findElement(By.className("content")).getText());

        driver.findElement(By.id("actionButton")).click();

        wait.until(ExpectedConditions.numberOfWindowsToBe(3));
        // Switch focus
        for(String handle : driver.getWindowHandles()) {
            driver.switchTo().window(handle);
        }

        wait.until(ExpectedConditions.visibilityOfElementLocated(By.className("content")));
        System.out.println("Current tab: " + driver.getWindowHandle());
         System.out.println("Page title: " + driver.getTitle());
        System.out.println(driver.findElement(By.className("content")).getText());
        driver.quit();

    }

    }
